### ISSUE TITLE
Publish public docker container

### ISSUE THREAD

**Issue by bentatham** [aerogear/keycloak-metrics-spi#34]
It would be great to have a published docker container to be used as an `init container` to keycloak kubernetes/helm installations, along with instructions to integrate that.

**Comment by grdryn** [comment]
@bentatham thanks for the suggestion, I think it's a good idea.

@pb82 We use this in an init container also, right? Is there anything that you could link here that shows how we do it, as an initial step?

**Comment by pb82** [comment]
@bentatham @grdryn Yes, we use an init container in the keycloak operator to install plugins (at the moment only the Metrics SPI).

The way it works is that we bundle the SPI together with a small Python script in a container named [sso plugins init](https://github.com/integr8ly/sso_plugins_init)

This container is then injected into the (upstream) sso template in the keycloak operator [here](https://github.com/integr8ly/keycloak-operator/blob/master/pkg/keycloak/install.go#L41).

We do this only because we don't want to modify and maintain our own sso template, so we inject our changes during runtime. If you have control over the template to deploy keycloak then you can just add the init container there.

When the init container runs it checks an env var named `SSO_PLUGINS` (see [here](https://github.com/integr8ly/sso_plugins_init/blob/master/plugins.py#L6)) which has the comma separated list of plugins. It then just copies the requested plugins into a volume mounted to the directory where the server expects extensions.

We have not formally documented this anywhere but i agree that it would make a good example use case.

**Comment by bentatham** [comment]
I'm trying to through this and understand.  Not 100% clear to me wit the details you gave.

We are deploying keycloak with helm charts (now at) codecentric/keycloak.

I built my own container with the keycloak-metrics-spi jar in it (FROM alpine)

I then used [these instructions](https://github.com/helm/charts/issues/4633) to mount the providers into `/opt/jboss/keycloak/providers`, which I can see.  However, adding `metrics-listener` to the config, I get this message:
```
20:08:23,589 ERROR [org.keycloak.services] (default task-1) KC-SERVICES0083: Event listener 'metrics-listener' registered, but provider not found
```

I suspect it is a permissions problem:  if I exec onto the pod, I can see:
```
[jboss@keycloak-id-dev-0 keycloak]$ ls -l /opt/jboss/keycloak/providers
total 88
-rw------- 1 root jboss 89622 Jun 25 19:56 keycloak-metrics-spi-1.0.1.jar
[jboss@keycloak-id-dev-0 keycloak]$ ls -l /opt/jboss/keycloak/
total 528
drwxrwx--x 3 jboss root    4096 Apr 23 14:09 bin
drwxrwxr-x 7 jboss root      92 Apr 23 14:09 docs
drwxrwx--x 5 jboss root      50 Apr 17 06:24 domain
-rw-rw-r-- 1 jboss root  488551 Apr 17 06:24 jboss-modules.jar
-rw-rw-r-- 1 jboss root   10637 Apr 17 06:24 License.html
-rw-rw-r-- 1 jboss root   26530 Apr 17 06:24 LICENSE.txt
drwxrwxr-x 3 jboss root      39 Apr 23 14:09 modules
drwxrwsrwx 2 root  jboss     44 Jun 25 19:56 providers
drwxrwx--x 1 jboss root      61 Apr 23 14:09 standalone
drwxrwxr-x 5 jboss root      76 Apr 23 14:08 themes
-rw-rw-r-- 1 jboss root      25 Mar  4 07:16 version.txt
drwxrwxr-x 2 jboss root      42 Apr 23 14:09 welcome-content
[jboss@keycloak-id-dev-0 keycloak]$ 
```

Any suggestions on how to fix that, of if permissions is indeed the issue?

**Comment by jkroepke** [comment]
I prefer to release pre build artifacts.

https://docs.travis-ci.com/user/deployment/releases/#draft-releases-with-draft-true

It's more flexible, since users just can run curl from github directly inside the init container, entrypoint script or just getting the aritfact inside the docker build if they build an own keycloak image.

**Comment by cleman95** [comment]
Hello, @bentatham I have the same issue. I mount the jar into my keycloak docker container at /opt/jboss/keycloak/providers and it returns the same error message as you. Did you fix the issue ? if yes, how ? thanks

**Comment by aqabawe** [comment]
@cleman95 looks like someone provided an answer to this here: https://github.com/helm/charts/issues/4633#issuecomment-509195839

@bentatham Did you manage to get this working? Can you share the image you created?

**Comment by eyalzek** [comment]
This is the `Dockerfile` I'm using and it works (versions probably outdated at this point):
```
FROM jboss/keycloak:9.0.0

# https://github.com/aerogear/keycloak-metrics-spi
ADD https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar \
    /opt/jboss/keycloak/providers/keycloak-metrics-spi-1.0.4.jar

USER root
RUN chown -R jboss:root /opt/jboss/keycloak/providers/ &&\
    chmod +x /opt/jboss/keycloak/providers/keycloak-metrics-spi-1.0.4.jar
USER jboss
```

**Comment by bogdanalov-sw** [comment]
> This is the `Dockerfile` I'm using and it works:
> 
> ```
> FROM jboss/keycloak:10.0.2
> 
> # https://github.com/aerogear/keycloak-metrics-spi
> ADD https://github.com/aerogear/keycloak-metrics-spi/releases/download/2.0.1/keycloak-metrics-spi-2.0.1.jar \
>     /opt/jboss/keycloak/providers/keycloak-metrics-spi-2.0.1.jar
> 
> USER root
> RUN chown -R jboss:root /opt/jboss/keycloak/providers/ &&\
>     chmod +x /opt/jboss/keycloak/providers/keycloak-metrics-spi-2.0.1.jar
> USER jboss
> ```

This didn't work for me. I had to add it to the "/opt/jboss/keycloak/standalone/deployments" and according to the latest documentation it should be there. Working solution for me is:



ARG KEYCLOAK_VERSION

FROM jboss/keycloak:$KEYCLOAK_VERSION

ADD https://github.com/aerogear/keycloak-metrics-spi/releases/download/2.0.1/keycloak-metrics-spi-2.0.1.jar \
    /opt/jboss/keycloak/standalone/deployments/keycloak-metrics-spi-2.0.1.jar

USER root
RUN chown -R jboss:root /opt/jboss/keycloak/standalone/deployments/ &&\
    chmod +x /opt/jboss/keycloak/standalone/deployments/keycloak-metrics-spi-2.0.1.jar &&\
    touch /opt/jboss/keycloak/standalone/deployments/keycloak-metrics-spi-2.0.2.jar.dodeploy 

USER jboss

**Comment by a72701** [comment]
If you're using the codecentric helm chart, you don't even need an init container or to build anything. Presume this would work fine for themes too

```
#values.yaml
extraVolumeMounts: | 
  - name: metrics-plugin
     mountPath: /opt/keycloak/providers/
extraVolumes: |
  - name: metrics-plugin
     secret:
     secretName: keycloak-metrics-plugin # this might need changing, I have 'keycloak' set as my full name override
secrets:
  metric-plugin:
     type: opaque
     data:
       keycloak-metrics-spi.jar: |
          <base64 data>
```
Then by running 
`cat keycloak-metrics-spi.jar | base64 >> values.yaml`
I can put a base64'd dump of the file into the secret. Its a large base64 dump, but allows me to mount the jar file directly in at runtime, and store the versioned jar file in my GitOps yaml files.

**Comment by james-callahan** [comment]
> Then by running
> `cat keycloak-metrics-spi.jar | base64 >> values.yaml`
> I can put a base64'd dump of the file into the secret. Its a large base64 dump, but allows me to mount the jar file directly in at runtime, and store the versioned jar file in my GitOps yaml files.

I'm amazed this doesn't exceed the maximum size for a secret.
The current release (3.0.0) is 151KB, which I guess is indeed smaller than the maximum size for a kubernetes secret.
Though I don't imagine its a great long-term solution.
